metrics.nim

```diff
@@ -221,6 +229,15 @@ template declarePublicCounter*(identifier: untyped,
   else:
     type identifier* = IgnoredCollector
 
+# alternative API
+proc counter*(name: string, labels: Labels = @[], registry = defaultRegistry): Counter =
```

Comment:
Why didn't you use the approach that I suggested (`statis string` + `{.global.}`)? it will avoid table look-ups on each counter usage.

When the metrics are disabled, you can rewrite the `counter("foo")` call to the `IgnoredCollector` symbol instead of returning `nil`

Reply:
I have to support the same collector being used in different files. How can I do that with global variables created in a template?
The problem is checking if a collector with that name already exists.

Reply:
The construct is this one:

```nim
proc counter*(name: static string): Counter {.inline.} =
  var c {.global.} = newCounter(name, "")
  return c
```

This work properly across multiple modules and it will be compiled to a function returning a known static address, which should be further optimized by the C compiler.

If I have to explain more, having a `static string` parameter makes the proc generic (it will be instantiated once per each unique value of the parameter). Each instantiation will create exactly 1 global location.

Reply:
But you're not supposed to create that collector on each function call. Only when it doesn't already exist.
Say you make the proc generic with static strings (weird hidden behaviour, BTW), what happens when you call `counter("foo")` from 2 different files?

Reply:
`.global.` variable are initialized once at program start-up. They are not even initialized lazily as in C++. Take a look at the generated code.

Reply:
This is weird. Works without `{.inline.}`, fails with it.
global.nim:
```nim
type Counter* = ref object of RootObj
  name*: string
  value*: int

proc `$`*(c: Counter): string =
  "name = " & c.name & ", value = " & $c.value

proc counter*(name: static string): Counter {.inline.} =
  var c {.global.} = Counter(name: name)
  return c

counter("foo").value = 1
```

global_main.nim:
```nim
import ./global

echo counter("foo") # the value is 0 instead of 1, with {.inline.}
counter("foo").value = 2
echo counter("foo")
```

Relevant C code being generated:
global.c:
```c
N_LIB_PRIVATE N_NIMCALL(void, unknown_globalInit000)(void) {
{
// [...]
	T100001_ = (tyObject_CountercolonObjectType__j9aUSVfSrFN8a9b2b9a4JP9cPg*) newObj((&NTI_nKkSDLQPZzH9a1MaaFn7DUw_), sizeof(tyObject_CountercolonObjectType__j9aUSVfSrFN8a9b2b9a4JP9cPg));
// [...]
	asgnRef((void**) (&c_lntSS3UjELzPLCdO49cM9bOw), T100001_);
```

global_main.c:
```c
N_LIB_PRIVATE N_NIMCALL(void, NimMainModule)(void) {
{
// [...]
	T100001_ = (tyObject_CountercolonObjectType__j9aUSVfSrFN8a9b2b9a4JP9cPg*) newObj((&NTI_nKkSDLQPZzH9a1MaaFn7DUw_), sizeof(tyObject_CountercolonObjectType__j9aUSVfSrFN8a9b2b9a4JP9cPg));
// [...]
	asgnRef((void**) (&c_lntSS3UjELzPLCdO49cM9bOw), T100001_);
```

Reply:
It gets worse, after adding this to global_main.nim:
```nim
echo counter("bar")
counter("bar").value = 3
echo counter("bar")
```

Output showing the new Counter object overwriting the old one:
```text
name = bar, value = 0
name = bar, value = 2
name = bar, value = 2
name = bar, value = 3
```
Also fixed by removing `{.inline.}`.

Reply:
Ah, right, sorry. Don't use inline, because this creates multiple definitions of the proc in the C code. With link-time code generation, I think the compiler will still inline it.

Reply:
sounds a bit like a nim codegen bug.. report upstream?

Reply:
Reported by @yglukhov, 2 years ago: https://github.com/nim-lang/Nim/issues/5945

Reply:
I replaced the hash table with {.global.} collector vars. Had to drop support for labels and custom registries, in the process.

Different collector types with the same name are now possible, due to {.global.} var name mangling rules.